### ISSUE TITLE
Boolean Operations via `three-csg`

### ISSUE THREAD

**Issue by ixxie** [threlte/threlte#443]
Implementing boolean operations - also known as Constructive Solid Geometry - is a common need in 3D applications.

Some library for doing this in Three.js called `CSG` has been floating around for years, and it seems to have been passed from one maintainer to another. The most recently maintained version I can find is a typescript port called [three-csg-ts](https://github.com/samalexander/three-csg-ts).

I propose to wrap this in some components to enable boolean operations:
```svelte
<Subtract>
  <T.BoxGeometry/>
  <T.SphereGeomtry/>
</Subtract>
```

Open questions:
  1. How should this be implemented?
  2. Which package should it go to (`@threlte/extras` or in a seperate `@threlte/csg` package)?

**Comment by jirihon** [comment]
I love this idea. You could describe solid models declaratively and easily connect them to user inputs.

I'd just like to add that there are other CSG engines out there. I'm convinced the current state of the art is [manifold](https://github.com/elalish/manifold). I'm using their webassembly release in my threlte project and it is really fast and capable. You can even preserve UV coordinates and normals throughout all the boolean operations, see [https://manifoldcad.org/model-viewer.html](https://manifoldcad.org/model-viewer.html). The only limitation is that manifold requires manifold meshes. But if you think a little big about CSG, the result for non-manifold meshes can't be well-defined.

**Comment by ixxie** [comment]
@jirihon thanks for the input! Would you be interested in submitting a PR for porting manifold to `@threlte/extras`?

**Comment by ixxie** [comment]
On second thought it looks like that is so big it would need to be its own package, something like `@threlte/manifold`

**Comment by michealparks** [comment]
Also want to mention that there's https://gkjohnson.github.io/three-bvh-csg/ as an option, which is used in the react-three-fiber ecosystem

**Comment by jirihon** [comment]
> @jirihon thanks for the input! Would you be interested in submitting a PR for porting manifold to `@threlte/extras`?

Oh, that would be way beyond my current abilities and current focus. I'm getting slowly into the magic of Threlte architecture and I am increasingly appreciating all the effort that has gone into it.

**Comment by ixxie** [comment]
@grischaerbe you had some thoughts about this; could you put them in writing to guide potential contributors?

**Comment by grischaerbe** [comment]
> @grischaerbe you had some thoughts about this; could you put them in writing to guide potential contributors?

Yes, of course. I do actually favor ["three-bvh-csg"](https://gkjohnson.github.io/three-bvh-csg/) because the API surface is just so much smaller and it seems feature complete enough to be a viable package in the Threlte ecosystem. The author of the package is known to deliver great stuff for Three.js already, so that's a plus as well.
